ingest-router/src/lib.rs

```diff
@@ -6,41 +6,124 @@ pub mod locale;
 pub mod project_config;
 pub mod router;
 
+#[cfg(test)]
+mod testutils;
+
+use crate::errors::IngestRouterError;
 use http_body_util::{BodyExt, Full, combinators::BoxBody};
+use hyper::StatusCode;
 use hyper::body::Bytes;
-use hyper::body::Incoming;
 use hyper::service::Service;
 use hyper::{Request, Response};
+use locator::client::Locator;
+use shared::http::make_error_response;
 use shared::http::run_http_service;
 use std::pin::Pin;
 
-#[derive(thiserror::Error, Debug)]
-pub enum IngestRouterError {
-    #[error("io error: {0}")]
-    Io(#[from] std::io::Error),
-}
-
 pub async fn run(config: config::Config) -> Result<(), IngestRouterError> {
-    let router_service = IngestRouterService {};
-    let router_task = run_http_service(&config.listener.host, config.listener.port, router_service);
+    let locator = Locator::new(config.locator.to_client_config()).await?;
+
+    let ingest_router_service = IngestRouterService {
+        router: router::Router::new(config.routes, locator),
+    };
+    let router_task = run_http_service(
+        &config.listener.host,
+        config.listener.port,
+        ingest_router_service,
+    );
     router_task.await?;
     Ok(())
 }
 
-struct IngestRouterService {}
+struct IngestRouterService {
+    router: router::Router,
+}
 
-impl Service<Request<Incoming>> for IngestRouterService {
+impl<B> Service<Request<B>> for IngestRouterService
```

Comment:
The generic request type `<B>` is used here instead of Incoming so we can test this API with synthetic requests. Same strategy as used by the proxy.